### PR TITLE
feature extraction, sentence similarity and text classification now use pipeline url

### PR DESCRIPTION
**Pull request by lisandropuzzolo**
Some models allows them to be use on more than one (default task).
In order to let the huggingface.js use more than the default model task, I've changed featureExtraction, sentenceSimilarity and textClassification to change the model to the pipeline url when model is not a url.

### CODE REVIEW

File: packages/inference/src/lib/makeRequestOptions.ts [hunk 1]
```diff
@@ -38,7 +48,7 @@ export function makeRequestOptions(
 		}
 	}
 
-	const url = /^http(s?):/.test(model) || model.startsWith("/") ? model : `${HF_INFERENCE_API_BASE_URL}${model}`;
+	const url = isModelTypeURL(model) ? model : `${HF_INFERENCE_API_MODEL_BASE_URL}${model}`;
```
Comment (coyotte508) [inline packages/inference/src/lib/makeRequestOptions.ts:51]: The `/^http(s?):/.test(model)` test needs to stay, it's for inference endpoints.

Only the latter part of the ternary expression needs to be changed

**Comment by coyotte508** [comment]
Thanks, that's a neat feature!

I would add a `pipeline` optional string param here: 
image

And do the logic of changing the url in `makeRequestOptions` instead.

In `textClassification` etc, it would be a one-line-change:

```ts
const res = (await request<TextClassificationOutput[]>(args, {
  ...options, 
  pipeline: "text-classification" // HERE!
}))?.[0];
```

**Review by coyotte508 — COMMENTED** [review]
Thanks, that's a neat feature!

I would add a `pipeline` optional string param here: 
image

And do the logic of changing the url in `makeRequestOptions` instead.

In `textClassification` etc, it would be a one-line-change:

```ts
const res = (await request<TextClassificationOutput[]>(args, {
  ...options, 
  pipeline: "text-classification" // HERE!
}))?.[0];
```

File: packages/inference/src/lib/makeRequestOptions.ts (hunk 1 above)
Comment (lisandropuzzolo) [reply packages/inference/src/lib/makeRequestOptions.ts:51]: I agree with the previous suggestions, but I'm not getting this one...
I've replaced the first part of the ternary expression with a a function call that just do the same:
```
export function isModelTypeURL(model: string): boolean {
	return /^http(s?):/.test(model) || model.startsWith("/");
}
```

I've done this way in order to not write the same expression in getPipelineURL(...)

Would you please clarify me?
Thanks in advance!

(I'm already implementing the other suggestions you made)

**Comment by lisandropuzzolo** [comment]
Done changes sugested by [coyotte508](https://github.com/coyotte508)

File: packages/inference/src/lib/makeRequestOptions.ts [hunk 2]
```diff
@@ -1,6 +1,20 @@
...
+	return /^http(s?):/.test(model) || model.startsWith("/");
+}
+
+function getPipelineURL(model: string, pipeline: string): string {
+	return `${HF_INFERENCE_API_PIPELINE_BASE_URL}${pipeline}/${model}`;
+}
+
+function getDefaultModelTaskURL(model: string) {
+	return `${HF_INFERENCE_API_MODEL_BASE_URL}${model}`;
+}
```
Comment (coyotte508) [inline packages/inference/src/lib/makeRequestOptions.ts:17]: Those are one-line local functions used once, can you hardcode them directly in `makeRequestOptions`?

File: packages/inference/src/tasks/nlp/featureExtraction.ts [hunk 3]
```diff
@@ -2,6 +2,8 @@ import { InferenceOutputError } from "../../lib/InferenceOutputError";
 import type { BaseArgs, Options } from "../../types";
 import { request } from "../custom/request";
 
+const HF_INFERENCE_API_PIPELINE_FEATURE_EXTRACTION = "feature-extraction";
```
Comment (coyotte508) [inline packages/inference/src/tasks/nlp/featureExtraction.ts:5]: The constant is only used once / not worth exporting, can you hardcode it below and remove the declaration here?

(maybe later we'll type `pipeline` as `"feature-extraction" | ... | string` to offer some kind of completion)

Same remark for the two other tasks

**Review by coyotte508 — COMMENTED** [review]
A few comments left, after they're handled I'll update `tapes.json` and merge !

**Review by julien-c — COMMENTED** [review]
would be interested in @Narsil (and @OlivierDehaene) feedback and maybe @Wauplin too on this

is this a feature of the Inference API we want to expose?

**Comment by lisandropuzzolo** [comment]
> would be interested in @Narsil (and @OlivierDehaene) feedback and maybe @Wauplin too on this
> 
> is this a feature of the Inference API we want to expose?

it's a very needed inference api feature, because many models can be used for more than one task

**Comment by Wauplin** [comment]
Makes me think about this recent convo (https://github.com/huggingface/huggingface_hub/pull/1474#discussion_r1196255344). In `huggingface_hub` we'll use the /task endpoint to allow this.

**Comment by coyotte508** [comment]
Sorry @lisandropuzzolo wrong git manipulation :grimacing:

**Pull request closed without merging**